<a href="https://colab.research.google.com/github/SofiMont1/Anal-tica-de-negocios/blob/main/ModeloArbol_Clase.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Arboles de Decisión**
0. Cargar las librerías de referencia para el modelo

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt #Graficador de cosas
from sklearn.tree import DecisionTreeClassifier

In [ ]:
#Para no tener que cargar el archivo
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


1. Se procede con la carga de los datos del archivo SolicitantesCredito

In [ ]:
nxl='/content/drive/MyDrive/Analítica de negocios/0. SolicitantesCrédito(USD) (2).xlsx'
XDB=pd.read_excel(nxl,sheet_name=1)
XDB=XDB.dropna() #Elimina celdas vacias
XDB.head(100)

#Se procede con la selección de las variables de trabajo
XD=np.array(XDB[['Edad', 'Ingresos', 'Egresos', 'Monto (EAD)']]) #Datos entrada
yd=np.array(XDB[['PreApr']]) #Este es el dato que quiero que el modelo pronóstique
print(yd)

[[1]
 [1]
 [0]
 ...
 [1]
 [1]
 [1]]


2. Creación del modelo - Decision Tree

In [ ]:
mar=DecisionTreeClassifier(criterion="gini", max_depth = 4) #4 niveles o variables de entrada
mar.fit(XD,yd)
ydp=mar.predict (XD)

from sklearn.metrics import confusion_matrix
cm = confusion_matrix(yd, ydp)
print(cm)

#Evaluamos el modelo
VN=cm [0,0]; FP=cm [0,1]; FN=cm [1,0]; VP=cm [1,1]

Ex=(VP+VN)/(VP+VN+FP+FN) #Que porcentaje de los datos pronóstico correctamente
print("La exactitud alcanzada por el modelo es",Ex)
Se=VP/(FN+VP) #Como le fue con los positivos
print("La sensibilidad alcanzada por el modelo es",Se)

[[2301  658]
 [ 644 2239]]
La exactitud alcanzada por el modelo es 0.7771311194796303
La sensibilidad alcanzada por el modelo es 0.7766215747485259


3. ¿Y dónde está est árbol?

In [ ]:
#Estas son las librearías para graficar el árbol
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs=['Edad', 'Ingresos', 'Egresos', 'Monto (EAD)'] #Son los títulos para los nodos
dot_graph=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_graph)
graph.write_png('arbol.png')

True

4.Se procede con el pronóstico de nuevos solicitantes

In [ ]:
XDB2=pd.read_excel(nxl,sheet_name=2)
XDB2=XDB2.dropna() #Elimina celdas vacias
XDB2.head(100)

XD2=np.array(XDB2[['Edad', 'Ingresos', 'Egresos', 'Monto (EAD)']]) #Datos entrada
print(XD2)

#Pronosticamos la PreAprobación
ydp2=mar.predict (XD2)
print(ydp2)
print("El número de PreAprobados es:",np.sum(ydp2))
print("El número total de datos es:",len(ydp2))
print("El número de PreNegados es:",len(ydp2)-np.sum(ydp2))

[[  38.          383.67775     368.59866667  175.84723507]
 [  47.           34.1875       38.03444444   26.95587069]
 [  46.         1774.305      2116.72377778  705.43470796]
 ...
 [  32.          283.26575     170.26644444  225.30034584]
 [  35.          768.288       847.26444444  479.16423193]
 [  24.          487.598       376.18355556  424.283721  ]]
[0 0 1 ... 0 1 0]
El número de PreAprobados es: 1260
El número total de datos es: 2536
El número de PreNegados es: 1276


**Anális de datos**
De acuerdo con la base de datos suministrada y el nodo base del árbol de decisión se puede observar un total de 5842 datos, divididos en 2959 PreAprobados y un total de 2883 PreNegados, esto representa el 51% y el 49% respectivamente. D eacuerdo con el pronóstco realizado por el modelo, este alcanzó una exactitud cercana al 77%, valores que fueorn similares a la Sensibilidad que mide el comportamiento del modelo frente a la calsificación de los PreAprobados, así como frente a la Especificidad, indicador que mide el comportamiento frente a la clasificación de los PreNegados.  Se destaca el nodo Puro "[PreNeg:10, PreApr:0]", en donde los solicitantes de crédito deben cumplir con la siguiente regla: SI Ing>644 AND Egr<=685 AND Ing>376 AND Monto>=322.
Con respecto al pronóstico, se puede observar un total de 2536 individuos sin PreAprobación, de los cuales un total de 1260 pueden ser PreAprobados y un total de 1276 que pueden ser PreNegados.